munit/shared/src/main/scala-2/munit/internal/MacroCompat.scala

```diff
@@ -1,14 +1,90 @@
 package munit.internal
 
 import munit.Clue
+import munit.Compare
 import munit.Location
 import scala.language.experimental.macros
 import scala.reflect.macros.blackbox.Context
 import scala.reflect.macros.TypecheckException
 import scala.reflect.macros.ParseException
+import scala.collection.mutable
 
 object MacroCompat {
 
+  trait CompareMacro {
+
+    /**
+     * Implements equality according to `==` for any two types (including
+     * unrelated types) as long as the compiler option
+     * "-Xmacro-settings:munit.strictEquality" is not enabled.
+     */
+    implicit def defaultCompareImplicit[A, B]: Compare[A, B] =
+      macro compareImpl[A, B]
+  }
+
+  private val strictEqualityFlag = "munit.strictEquality"
+  private var strictEqualityCache: Option[Boolean] = None
+  private var strictEqualityError: Option[Context => Unit] = None
+  private def isStrictEqualityEnabled(c: Context): Boolean = {
+    strictEqualityCache match {
+      case Some(cachedValue) =>
+        strictEqualityError.foreach(fn => fn(c))
+        cachedValue
+      case _ =>
+        val result = c.settings.contains(strictEqualityFlag)
+        c.settings.foreach { setting =>
+          if (setting.startsWith("munit.") && setting != strictEqualityFlag) {
+            strictEqualityError = Some(context =>
+              context.error(
+                context.enclosingPosition,
+                s"unknown flag '-Xmacro-setting:$setting'. Did you mean -Xmacro-setting:$strictEqualityFlag?"
+              )
+            )
+          }
+        }
+        strictEqualityCache = Some(result)
+        isStrictEqualityEnabled(c)
+    }
+  }
+
+  def compareImpl[A: c.WeakTypeTag, B: c.WeakTypeTag](
+      c: Context
+  ): c.Expr[Compare[A, B]] = {
+    import c.universe._
+    if (!isStrictEqualityEnabled(c)) {
+      reify(Compare.defaultCompare[A, B])
+    } else {
```

Comment:
minor concern: this doesn't cover the case in which a user wants to customize the diff but use universal equality. Right now it's not a big issue, but it may not scale if we make more customizations available via the `Compare` typeclass

Reply:
If you have a custom implicit `Compare[A, B]` in scope then you won't hit on this macro in the first place. My first attempt to implement this macro had infinite recursion because I tried to fallback to `implicitly[Compare[A, B]]`, which triggered another macro expansion 😅 There's no way to do "pick the next implicit in scope", as far as I understand